pkg/controller/pxc/replication.go

```diff
@@ -86,6 +96,291 @@ func (r *ReconcilePerconaXtraDBCluster) removeOutdatedServices(cr *api.PerconaXt
+	if err != nil {
+		return errors.Wrap(err, "get primary pxc pod")
+	}
+
+	var primaryPod *corev1.Pod
+	for _, pod := range list.Items {
+		if pod.Status.PodIP == primary || pod.Name == primary || strings.HasPrefix(primary, fmt.Sprintf("%s.%s.%s", pod.Name, sfs.StatefulSet().Name, cr.Namespace)) {
+			primaryPod = &pod
+			break
+		}
+	}
+
+	if primaryPod == nil {
+		logger.Info("Unable to find primary pod for replication. No pod with name or ip like this", "primary name", primary)
+		return nil
+	}
+
+	user := "operator"
+	port := int32(33062)
+
+	primaryDB, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, primaryPod.Name+"."+cr.Name+"-pxc."+cr.Namespace, port)
+	if err != nil {
+		return errors.Wrapf(err, "failed to connect to pod %s", primaryPod.Name)
+	}
+
+	defer primaryDB.Close()
+
+	dbVer, err := primaryDB.Version()
+	if err != nil {
+		return errors.Wrap(err, "failed to get current db version")
+	}
+
+	if minReplicationVersion.Compare(version.Must(version.NewVersion(dbVer))) < 0 {
+		logger.Info("Failed to reconcile replication. Unsupported mysql version", "minimum version", minReplicationVersion.String())
+		return nil
+	}
+
+	err = removeOutdatedChannels(primaryDB, cr.Spec.PXC.ReplicationChannels)
+	if err != nil {
+		return errors.Wrap(err, "remove outdated replication channels")
+	}
+
+	err = checkReadonlyStatus(cr.Spec.PXC.ReplicationChannels, list.Items, cr, r.client)
+	if err != nil {
+		return errors.Wrap(err, "failed to ensure cluster readonly status")
+	}
+
+	if len(cr.Spec.PXC.ReplicationChannels) == 0 {
+		return nil
+	}
+
+	// if primary pod is not a replica, we need to make it as replica, and stop replication on other pods
+	_, ok := primaryPod.Labels[replicationPodLabel]
+	if !ok {
+		for _, pod := range list.Items {
+			if pod.Name == primaryPod.Name {
+				continue
+			}
+			if _, ok := pod.Labels[replicationPodLabel]; ok {
+				r.logger(cr.Name, cr.Namespace).Info("Replication pod has changed", "new replication pod", pod.Name)
+				db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, pod.Name+"."+cr.Name+"-pxc."+cr.Namespace, port)
+				if err != nil {
+					return errors.Wrapf(err, "failed to connect to pod %s", pod.Name)
+				}
+				err = db.StopAllReplication()
+				db.Close()
+				if err != nil {
+					return errors.Wrapf(err, "stop replication on pod %s", pod.Name)
+				}
+				delete(pod.Labels, replicationPodLabel)
+				err = r.client.Update(context.TODO(), &pod)
+				if err != nil {
+					return errors.Wrap(err, "failed to remove primary label from secondary pod")
+				}
+			}
+		}
+		primaryPod.Labels[replicationPodLabel] = "true"
+		err = r.client.Update(context.TODO(), primaryPod)
+		if err != nil {
+			return errors.Wrap(err, "add label to main replica pod")
+		}
+	}
+
+	sysUsersSecretObj := corev1.Secret{}
+	err = r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Spec.SecretsName,
```

Comment:
Should we use internal secret for these needs?   I think yes.

Reply:
eedc24c